### DISCUSSION TITLE
error when use ActivateElementsByPath and mesh adaptivity at the same time

### DISCUSSION THREAD

**Discussion by cztkatherine** [idaholab/moose#18284] [Q&A Modules: General] [answered]
Dear colleagues,

It appears that when we use ActivateElementsByPath and mesh adaptivity at the same time, there would be an error saying:

YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)

We are able to reproduce this error by adapting the example (modules/combined/test/tests/additive_manufacturing/check_element_addition_2D.i), and a copy of the modified input is attached.
```
[Problem]
  kernel_coverage_check = false
[]

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    ymin = 0
    xmax = 1
    ymax = 0.5
    nx = 10
    ny = 10
  []
  [bottom_domain]
    input = gen
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0 0'
    top_right = ' 1 0.1 0'
    block_id = 1
  []
  [top_domain]
    input = bottom_domain
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0.1 0'
    top_right = '1 0.5 00'
    block_id = 2
  []
  [sidesets]
    input = top_domain
    type = SideSetsAroundSubdomainGenerator
    normal = '1 0 0'
    block = 1
    new_boundary = 'moving_interface'
  []
[]

[Variables]
  [temp]
    block = '1 2'
    initial_condition = '20'
  []
[]

[Functions]
  [fy]
    type = ParsedFunction
    value = '0.2'
  []
  [fx]
    type = ParsedFunction
    value = 't'
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  automatic_scaling = true
  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  line_search = 'none'

  l_max_its = 10
  nl_max_its = 20
  nl_rel_tol = 1e-4

  start_time = 0.0
  end_time = 1
  dt = 5e-2
  dtmin = 1e-4
[]


[Adaptivity]
  marker = Marker3 #this line when commented, switches off adaptivity
  max_h_level = 3
  #steps = 2 #this line gets ignored in a transient run
  [./Indicators]
    [./Terror]
      type = GradientJumpIndicator
      variable = temp
      outputs = none
    [../]   
  [../]
  [./Markers]
    [./Marker3]
      type = ErrorFractionMarker
      refine = 0.65
      coarsen = 0.25
      indicator = Terror
      outputs = none
    [../]
  [../]
[]


[UserObjects]
  [activated_elem_uo]
    type = ActivateElementsByPath
    execute_on = timestep_begin
    activate_distance = 0.2
    function_x = fx
    function_y = fy
    active_subdomain_id = 1
    expand_boundary_name = 'moving_interface'
  []
[]

[Outputs]
  exodus = true
file_base = ./CheckEleAddition2D/CheckEleAddition2D
[]
```
When we run with the adaptivity block, the above error appears. If you delete the adaptivity block, the simulation runs successfully.

Could any one share some thoughts about this conflicts?

Thank you.

**Comment by jiangwen84** [comment]
That might be a corner case that has not been fully tested.  @dewenyushu do you have any thoughts?

**Comment by hugary1995** [comment, marked as answer]
We plan to replace `ActivateElementsByPath` with the new element subdomain modifiers. I modified your input file to use the `CoupledVarThresholdElementSubdomainModifier` and it works fine with adaptivity. 

Here is the animation:
https://user-images.githubusercontent.com/31555204/124855357-d5246800-df76-11eb-9ffe-a9d563d93df7.mov

The modified input file is attached below.
```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    ymin = 0
    xmax = 1
    ymax = 0.5
    nx = 10
    ny = 10
  []
  [bottom_domain]
    input = gen
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0 0'
    top_right = ' 1 0.1 0'
    block_id = 1
  []
  [top_domain]
    input = bottom_domain
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0.1 0'
    top_right = '1 0.5 00'
    block_id = 2
  []
  [sidesets]
    input = top_domain
    type = SideSetsAroundSubdomainGenerator
    normal = '1 0 0'
    block = 1
    new_boundary = 'moving_interface'
  []
[]

[Variables]
  [temp]
    block = '1 2'
    initial_condition = '20'
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = temp
  []
[]

[BCs]
  [left]
    type = DirichletBC
    variable = temp
    value = 20
    boundary = 'left'
  []
  [right]
    type = FunctionDirichletBC
    variable = temp
    function = '20+10*t'
    boundary = right
  []
[]

[Functions]
  [fy]
    type = ParsedFunction
    value = '0.2'
  []
  [fx]
    type = ParsedFunction
    value = 't'
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  automatic_scaling = true
  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  line_search = 'none'

  l_max_its = 10
  nl_max_its = 20
  nl_rel_tol = 1e-4

  start_time = 0.0
  end_time = 1
  dt = 5e-2
  dtmin = 1e-4
[]

[Adaptivity]
  marker = threshold
  max_h_level = 1
  [Markers]
    [threshold]
      type = ValueThresholdMarker
      variable = temp
      refine = 22
    []
  []
[]

[AuxVariables]
  [dist]
  []
[]

[AuxKernels]
  [dist]
    type = ParsedAux
    variable = dist
    function = 'sqrt((x-t)^2+(y-0.2)^2)'
    use_xyzt = true
    execute_on = 'TIMESTEP_BEGIN LINEAR TIMESTEP_END'
  []
[]

[UserObjects]
  [esm]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = dist
    subdomain_id = 1
    threshold = 0.2
    criterion_type = BELOW
    apply_initial_conditions = false
    moving_boundary_name = 'moving_interface'
    block = 2
    execute_on = 'TIMESTEP_BEGIN'
  []
[]

[Outputs]
  exodus = true
  file_base = ./CheckEleAddition2D/CheckEleAddition2D
[]
```

**Reply by hugary1995** [reply]
But we should definitely add some regression tests including adaptivity.

**Comment by dewenyushu** [comment]
@jiangwen84 you're right. The `ActivateElementsByPath` has not been tested with mesh adaptivity.

@hugary1995 's suggestion is the right way to go.

**Reply by zx1987** [reply]
Thank you all @jiangwen84 @hugary1995 @dewenyushu and we will be using `CoupledVarThresholdElementSubdomainModifier.` 
Best,
Xiang

**Reply by zx1987** [reply]
I just have a quick test, it appears when adaptivity is based on `ValueThresholdMarker`, both `CoupledVarThresholdElementSubdomainModifier` and `ActivateElementsByPath` will work. If we use `GradientJumpIndicator`, neither of them will work. To be more specific, if we replace the adaptivity block of the input file provide by @hugary1995 using the block below:
[Adaptivity]
  marker = Marker3 #this line when commented, switches off adaptivity
  max_h_level = 3
  [./Indicators]
    [./Terror]
      type = GradientJumpIndicator
      variable = temp
      outputs = none
    [../]   
  [../]
  [./Markers]
    [./Marker3]
      type = ErrorFractionMarker
      refine = 0.65
      coarsen = 0.25
      indicator = Terror
      outputs = none
    [../]
  [../]
[]

The same error will occur. Could any of you take another look at it? Thank you all. 

Best,
Xiang

**Reply by hugary1995** [reply]
This marker doesn't make sense in this problem -- the gradient jump is zero everywhere.

**Reply by hugary1995** [reply]
Well, it probably implies that there are still some issues when combining ESMs (element subdomain modifier) with AMR (adaptive mesh refinement). If there is a more concrete non-working example, feel free to open an issue.

**Reply by zx1987** [reply]
You are right @hugary1995 about zero temperature gradient in the above input file. I slightly modified your input file as below to change the left BC to a higher temperature to introduce the temperature gradient and same error persists. If switching to the `ValueThresholdMarker`, then it behaves as expected. 

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    ymin = 0
    xmax = 1
    ymax = 0.5
    nx = 10
    ny = 10
  []
  [bottom_domain]
    input = gen
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0 0'
    top_right = ' 1 0.1 0'
    block_id = 1
  []
  [top_domain]
    input = bottom_domain
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0.1 0'
    top_right = '1 0.5 00'
    block_id = 2
  []
  [sidesets]
    input = top_domain
    type = SideSetsAroundSubdomainGenerator
    normal = '1 0 0'
    block = 1
    new_boundary = 'moving_interface'
  []
[]

[Variables]
  [temp]
    block = '1 2'
    initial_condition = '20'
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = temp
  []
[]

[BCs]
  [left]
    type = DirichletBC
    variable = temp
    value = 50
    boundary = 'left'
  []
  [right]
    type = FunctionDirichletBC
    variable = temp
    function = '20+10*t'
    boundary = right
  []
[]

[Functions]
  [fy]
    type = ParsedFunction
    value = '0.2'
  []
  [fx]
    type = ParsedFunction
    value = 't'
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  automatic_scaling = true
  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  line_search = 'none'

  l_max_its = 10
  nl_max_its = 20
  nl_rel_tol = 1e-4

  start_time = 0.0
  end_time = 1
  dt = 5e-2
  dtmin = 1e-4
[]

#[Adaptivity]
#marker = threshold
#max_h_level = 1
#[Markers]
#[threshold]
#type = ValueThresholdMarker
#variable = temp
#refine = 40
#[]
#[]
#[]


[Adaptivity]
  marker = Marker3 #this line when commented, switches off adaptivity
  max_h_level = 3
  [./Indicators]
    [./Terror]
      type = GradientJumpIndicator
      variable = temp
      outputs = none
    [../]   
  [../]
  [./Markers]
    [./Marker3]
      type = ErrorFractionMarker
      refine = 0.65
      coarsen = 0.25
      indicator = Terror
      outputs = none
    [../]
  [../]
[]

[AuxVariables]
  [dist]
  []
[]

[AuxKernels]
  [dist]
    type = ParsedAux
    variable = dist
    function = 'sqrt((x-t)^2+(y-0.2)^2)'
    use_xyzt = true
    execute_on = 'TIMESTEP_BEGIN LINEAR TIMESTEP_END'
  []
[]

[UserObjects]
  [esm]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = dist
    subdomain_id = 1
    threshold = 0.2
    criterion_type = BELOW
    apply_initial_conditions = false
    moving_boundary_name = 'moving_interface'
    block = 2
    execute_on = 'TIMESTEP_BEGIN'
  []
[]

[Outputs]
  exodus = true
  file_base = ./CheckEleAddition2D/CheckEleAddition2D
[]

**Reply by hugary1995** [reply]
Sorry I didn't follow. What did you do to make the temperature gradient jump nonzero?

**Reply by zx1987** [reply]
Sorry not making it clear. I simply change the prescribed temperature at the left boundary from 20 to 50. A snapshot of the temperature contour shows the expected temperature gradient if you run without adaptivity or using `ValueThresholdMarker`. 

![image](https://user-images.githubusercontent.com/11442693/124959838-e0a87b00-dfd8-11eb-9efe-8c69f4544b84.png)


Best,
Xiang

**Reply by zx1987** [reply]
Sorry I think I made a false statement here. The temperature is linearly varying, and the gradient is constant. I will further work on that and post it here.

**Reply by zx1987** [reply]
Hi Gary,
The smallest example I can replicate the error I have regarding `GradientJumpIndicator` based mesh adaptivity  and `CoupledVarThresholdElementSubdomainModifier` is by combining the example we have with the `FunctionPathEllipsoidHeatSource`. A copy of the input file is shown below. The simulation will terminate after step 9 due to segmentation fault. Any suggestion is appreciated. Thank you.

Best,
Xiang

[Mesh]
  [gen]
  type = GeneratedMeshGenerator
  dim = 3
  xmin = -5.0
  xmax = 5.0
  nx = 20

  ymin = -5.0
  ymax = 5.0
  ny = 20

  zmin = 0.0
  zmax = 1.0
  nz = 1
  []

  [left_domain]
    input = gen
    type = SubdomainBoundingBoxGenerator
    bottom_left = '-5 -5 0'
    top_right = ' 0 5 1'
    block_id = 1
  []
  [right_domain]
    input = left_domain
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 -5 0'
    top_right = '5 5  1'
    block_id = 2
  []
  [sidesets]
    input = right_domain
    type = SideSetsAroundSubdomainGenerator
    normal = '1 0  0'
    block = 1
    new_boundary = 'moving_interface'
  []

[]

[Variables]
  [./temp]
    initial_condition = 30
  [../]
[]

[Kernels]
  [./time]
    type = ADHeatConductionTimeDerivative
    variable = temp
  [../]
  [./heat_conduct]
    type = ADHeatConduction
    variable = temp
    thermal_conductivity = thermal_conductivity
  [../]
  [./heat_source]
    type = ADMatHeatSource
    material_property = volumetric_heat
    variable = temp
  [../]
[]

[BCs]
  [./temp_bottom_fix]
    type = ADDirichletBC
    variable = temp
    boundary = 1
    value = 30
  [../]
[]

[Materials]
  [./heat]
    type = ADHeatConductionMaterial
    specific_heat = 603
    thermal_conductivity = 10e-2
  [../]
  [./density]
    type = ADGenericConstantMaterial
    prop_names = 'density'
    prop_values = '4.43e-6'
  [../]
  [./volumetric_heat]
    type = FunctionPathEllipsoidHeatSource
    rx = 1
    ry = 1
    rz = 1
    power = 100
    efficiency = 0.5
    factor = 2
    function_x= path_x
    function_y= path_y
    function_z= path_z
  [../]
[]

[Functions]
  [./path_x]
    type = ParsedFunction
    #value = 2*cos(2.0*pi*t)
    value = 2*cos(2.0*pi*t)
  [../]
  [./path_y]
    type = ParsedFunction
    value = 2*sin(2.0*pi*t)
  [../]
  [./path_z]
    type = ParsedFunction
    value = 1.0
  [../]
[]


[AuxVariables]
  [dist]
  []
[]

[AuxKernels]
  [dist]
    type = ParsedAux
    variable = dist
    function = 'sqrt((y+5-t*10)^2+(x-1.5)^2+(z-0.5)^2)'
    use_xyzt = true
    execute_on = 'TIMESTEP_BEGIN LINEAR TIMESTEP_END'
  []
[]

[UserObjects]
  [esm]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = dist
    subdomain_id = 1    #--- new ID
    threshold = 1.5
    criterion_type = BELOW
    apply_initial_conditions = false
    moving_boundary_name = 'moving_interface'
    block = 2          #-- old id
    execute_on = 'TIMESTEP_BEGIN'
  []
[]

[Postprocessors]
  [temp_max]
    type = ElementExtremeValue
    variable = temp
  []
  [temp_min]
    type = ElementExtremeValue
    variable = temp
    value_type = min
  []
  [temp_avg]
    type = ElementAverageValue
    variable = temp
  []
[]

[Preconditioning]
  [./full]
    type = SMP
    full = true
  [../]
[]

[Adaptivity]
  marker = Marker3 #this line when commented, switches off adaptivity
  max_h_level = 3
  [./Indicators]
    [./Terror]
      type = GradientJumpIndicator
      variable = temp
      outputs = none
    [../]   
  [../]
  [./Markers]
    [./Marker3]
      type = ErrorFractionMarker
      refine = 0.75
      coarsen = 0.25
      indicator = Terror
      outputs = none
    [../]
  [../]
[]


[Executioner]
  type = Transient
  solve_type = PJFNK

  nl_rel_tol = 1e-6
  nl_abs_tol = 1e-6

  petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'preonly lu       superlu_dist'

  l_max_its = 100

  end_time = 1
  dt = 0.05
  dtmin = 1e-4
[]

[Outputs]
  csv = true
  exodus = true
  file_base = ./FunctionHeatSource/FunctionHeatSource
[]

**Reply by hugary1995** [reply]
Thanks for reporting that. I've created an issue #18368

**Reply by hugary1995** [reply]
The issue is that we are trying to remove an element side from the `moving_interface` sideset that belongs to a children element (created by adaptivity) that no longer exists. I still need to think about how best to fix this, but for now you can work around it by removing the parameter `moving_boundary_name = 'moving_interface'`, and the element subdomain modifier will not try to modify the sideset.

See the animation below, it work well.

https://user-images.githubusercontent.com/31555204/131875632-a529d942-04d7-4a35-91e7-69e3447cd91b.mp4

**Reply by zx1987** [reply]
Thank you Gary for letting me know. Yes, this works, and we have being using a fine mesh without refinement. I will check back when you finalize the fix. Thank you.

Best,
Xiang

**Reply by doquang** [reply]
@hugary1995 - Could you have a look in this example. The option `moving_boundary_name` was not used in this case, but the model still stop at step 2.
```
Time Step 2, time = 2, dt = 1
Assertion `my_dof_g == their_dof_g' failed.
my_dof_g = 3
their_dof_g = 1152921504606846976


Stack frames: 27
0: 0   libmesh_dbg.0.dylib                 0x000000010e00e180 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 488
1: 1   libmesh_dbg.0.dylib                 0x000000010dff0510 libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 248
2: 2   libmesh_dbg.0.dylib                 0x000000010e3d9fb0 libMesh::(anonymous namespace)::lagrange_compute_constraints(libMesh::DofConstraints&, libMesh::DofMap&, unsigned int, libMesh::Elem const*, unsigned int) + 4756
3: 3   libmesh_dbg.0.dylib                 0x000000010e3d8d10 libMesh::FE<2u, (libMesh::FEFamily)0>::compute_constraints(libMesh::DofConstraints&, libMesh::DofMap&, unsigned int, libMesh::Elem const*) + 52
4: 4   libmesh_dbg.0.dylib                 0x000000010e3bd5bc libMesh::FEInterface::compute_constraints(libMesh::DofConstraints&, libMesh::DofMap&, unsigned int, libMesh::Elem const*) + 644
5: 5   libmesh_dbg.0.dylib                 0x000000010df36064 (anonymous namespace)::ComputeConstraints::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&) const + 368
6: 6   libmesh_dbg.0.dylib                 0x000000010df08b20 void libMesh::Threads::parallel_for<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, (anonymous namespace)::ComputeConstraints>(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, (anonymous namespace)::ComputeConstraints const&) + 88
7: 7   libmesh_dbg.0.dylib                 0x000000010df0868c libMesh::DofMap::create_dof_constraints(libMesh::MeshBase const&, double) + 2296
8: 8   libmesh_dbg.0.dylib                 0x00000001111ea6f0 libMesh::System::reinit_constraints() + 48
9: 9   libmesh_dbg.0.dylib                 0x0000000111136394 libMesh::EquationSystems::reinit_solutions() + 2024
10: 10  libmesh_dbg.0.dylib                 0x0000000111135b78 libMesh::EquationSystems::reinit() + 32
``` 
```

[Problem]
  kernel_coverage_check = false
[]

[Mesh]
  [mesh1]
     type = GeneratedMeshGenerator
     dim = 2
     nx = 2
     ny = 2
     xmax = 2
     ymax = 2
  []
  [bottom_domain]
    input = mesh1
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0 0'
    top_right = '2 1 0'
    block_id = 1
  []
  [top_domain]
    input = bottom_domain
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 1 0'
    top_right = '2 2 0'
    block_id = 2
  []
[]

[Variables]
  [temp]
    block = '1'
    initial_condition = '1'
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
    [path_var_solid]
    []
[]

[AuxKernels]
    [path_elem_solid]
        type = ParsedAux
        variable = path_var_solid
        function = 'y+t'
        use_xyzt = true
        execute_on ='INITIAL TIMESTEP_BEGIN'
    []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  solve_type = PJFNK

  nl_rel_tol = 1e-6
  nl_abs_tol = 1e-6

  petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'preonly lu superlu_dist'

  l_max_its = 100

  start_time = 0.0
  end_time = 3
  dt = 1
  dtmin = 1e-4
[]

[Adaptivity]
  marker = marker
  max_h_level = 1
  [Markers]
        [marker]
            type = BoxMarker
            bottom_left = '0 0 0'
            top_right = '1 2 0'
            inside = refine
            outside = do_nothing
        []
  []
[]

[UserObjects]
    [activated_elem_solid]
        type = CoupledVarThresholdElementSubdomainModifier
        coupled_var = path_var_solid
        criterion_type = ABOVE
        threshold = 3
        subdomain_id = 1
        execute_on = 'TIMESTEP_BEGIN'
        apply_initial_conditions = true
    []
[]

[Outputs]
  exodus = true
  file_base = ./CheckEleAddition2D/CheckEleAddition2D
[]
```

**Reply by hugary1995** [reply]
@doquang Thanks for reporting! I can confirm this bug. I've created a ticket here #18835. The backtrace suggests that it segfaults while add constraint dofs in the equation system, but it seems to me that the input file doesn't have any constraints... I'll have to ask @roystgnr about it.

**Reply by roystgnr** [reply]
> it seems to me that the input file doesn't have any constraints

You don't need constraints in an input file to end up with constraints in your system; all you need is to refine one element while leaving one neighbor unrefined, and if you have a C0 or C1 FE defined on that element then that's it: libMesh has to generate at least one constraint equation to maintain that continuity level at the "hanging node".

What's going on here, though I don't know.  their_dof_g looks like it's obviously corrupted, but we just get that from examining DoF indices on one side of a parent of an active element.  Is it possible that you've deleted one or more child elements of a parent element without actually coarsening to make the parent active?  libMesh doesn't support that.
